Lib/shutil.py

```diff
@@ -764,43 +741,16 @@ def onexc(*args):
             path = os.fsdecode(path)
         # Note: To guard against symlink races, we use the standard
         # lstat()/open()/fstat() trick.
+        stack = [(os.lstat, dir_fd, path, None)]
         try:
-            orig_st = os.lstat(path, dir_fd=dir_fd)
-        except OSError as err:
-            onexc(os.lstat, path, err)
-            return
-        try:
-            fd = os.open(path, os.O_RDONLY | os.O_NONBLOCK, dir_fd=dir_fd)
-            fd_closed = False
-        except OSError as err:
-            onexc(os.open, path, err)
-            return
-        try:
-            if os.path.samestat(orig_st, os.fstat(fd)):
-                _rmtree_safe_fd(fd, path, onexc)
-                try:
-                    os.close(fd)
-                except OSError as err:
-                    # close() should not be retried after an error.
-                    fd_closed = True
-                    onexc(os.close, path, err)
-                fd_closed = True
-                try:
-                    os.rmdir(path, dir_fd=dir_fd)
-                except OSError as err:
-                    onexc(os.rmdir, path, err)
-            else:
-                try:
-                    # symlinks to directories are forbidden, see bug #1669
-                    raise OSError("Cannot call rmtree on a symbolic link")
-                except OSError as err:
-                    onexc(os.path.islink, path, err)
+            while stack:
+                _rmtree_safe_fd(stack, onexc)
         finally:
-            if not fd_closed:
-                try:
+            # Close any file descriptors still on the stack.
+            while stack:
+                func, fd, path, entry = stack.pop()
+                if func is os.close:
                     os.close(fd)
```

Comment:
It reads a bit weird to me to not use func in the call after, but that's a nit.

More importantly, previously the code was always guarding against `os.close` failing. Are you sure we no longer need that?

Reply:
I've adjusted the code to catch and report errors from this `os.close()` now.

The `func` is used to indicate what operation to perform - `os.lstat` kicks off a walk, `os.rmdir` removes a directory, and `os.close` closes a file descriptor. This `finally` block only cleans up file descriptors, so it needs to ignore the non-`os.close` stack items.